cloudinit/config/cc_ansible.py

```diff
@@ -0,0 +1,187 @@
+        return None
+
+    def pull(self, *args) -> str:
+        stdout, _ = subp([*self.cmd_pull, *args], env=self.env)
+        return stdout
+
+    def check_deps(self):
+        if not self.is_installed():
+            raise ValueError("command: ansible is not installed")
+
+    @abc.abstractmethod
+    def is_installed(self):
+        pass
+
+    @abc.abstractmethod
+    def install(self, pkg_name: str):
+        pass
+
+
+class AnsiblePullPip(AnsiblePull):
+    def __init__(self):
+        self.cmd_pull = ["ansible-pull"]
+        self.cmd_version = ["ansible-pull", "--version"]
+        self.env["PATH"] = ":".join([self.env["PATH"], "/root/.local/bin/"])
+
+    def install(self, pkg_name: str):
+        """should cloud-init grow an interface for non-distro package
+        managers? this seems reusable
+        """
+        if not self.is_installed():
+            subp(["python3", "-m", "pip", "install", "--user", pkg_name])
+
+    def is_installed(self) -> bool:
+        stdout, _ = subp(["python3", "-m", "pip", "list"])
+        return "ansible" in stdout
+
+
+class AnsiblePullDistro(AnsiblePull):
+    def __init__(self, distro):
+        self.cmd_pull = ["ansible-pull"]
+        self.cmd_version = ["ansible-pull", "--version"]
+        self.distro = distro
+
+    def install(self, pkg_name: str):
+        if not self.is_installed():
+            self.distro.install_packages(pkg_name)
+
+    def is_installed(self) -> bool:
+        return bool(which("ansible"))
+
+
+def handle(name: str, cfg: dict, cloud: Cloud, log: Logger, _):
+    ansible_cfg: dict = cfg.get("ansible", {})
+    if ansible_cfg:
+        validate_config(ansible_cfg)
+        install = ansible_cfg["install-method"]
+        pull_cfg = ansible_cfg.get("pull")
+        if pull_cfg:
+            ansible: AnsiblePull
+            if install == "pip":
+                ansible = AnsiblePullPip()
+            else:
+                ansible = AnsiblePullDistro(cloud.distro)
+            ansible.install(ansible_cfg["package-name"])
+            ansible.check_deps()
+            run_ansible_pull(ansible, deepcopy(pull_cfg), log)
+
+
+def validate_config(cfg: dict):
+    required_keys = {
+        "install-method",
+        "package-name",
+        "pull/url",
+        "pull/playbook-name",
+    }
+    for key in required_keys:
+        if not get_cfg_by_path(cfg, key):
+            raise ValueError(f"Invalid value config key: '{key}'")
+
+    install = cfg["install-method"]
+    if install not in ("pip", "distro"):
+        raise ValueError("Invalid install method {install}")
+
+
+def filter_args(cfg: dict) -> dict:
+    """remove boolean false values"""
+    return {key: value for (key, value) in cfg.items() if value is not False}
+
+
+def run_ansible_pull(pull: AnsiblePull, cfg: dict, log: Logger):
+    playbook_name: str = cfg.pop("playbook-name")
+
+    v = pull.get_version()
+    if not v:
+        log.warn("Cannot parse ansible version")
```

Comment:
We probably want to avoid using the log param from handle(.., log) in general it a design approach  that is antiquated and should eventually be retired from our `handle()` signatures. I added ticket `SC-1219` to capture that tech debt. Instead we want to use module-level loggers `LOG = logging.getLogger(__name__)` which give us a good prefix for the module generating the logs and avoid passing a log param into every separate function which needs it.